### DISCUSSION TITLE
Python Integration?

### DISCUSSION THREAD

**Discussion by heezes** [rokath/trice#263] [Q&A] [answered]
Greetings, 
First off the projects seems great. I plan on using it. Just a quick question I am currently using a setup in which i need to analyze my device trace via python, so is there any integration of python with trice available. Preferably just after Uart or RTT, if so please point me towards the appropriate docs.
Regards,
Altamash

**Comment by rokath** [comment]
Never used Python but it should be no big deal for you. I see these options:

1. Read https://github.com/rokath/trice/blob/master/docs/TriceMessagesEncoding.md and https://pythonhosted.org/cobs/cobs.cobs.html. 
  - Write a Python program to read til.json
  - Extract Id and values from binary data, get format string from til.json and printf 
2. Start `trice log -p COM1` as process from Python and grab the output.
3. Start `trice log -p COM1 -ds` and receive the emitted lines over TCP in a Python program.
  - The lines are transferred as Go string slices and it should  be possible to grab the strings in Python. 
  - I could provide an additional CLI switch to transfer pure strings to TCP.

**Reply by heezes** [reply]
Currently I'm using the 3rd method. Although now I'm tempted to explore the decoding and print process as mentioned above.
Btw Thanks for the help

**Reply by watersb1323** [reply]
Hi @heezes, just wondering did you give processing the raw data with Python (option 1) a go? I think this would be a good idea for a repo, looking at implementing something similar myself.

**Reply by heezes** [reply]
So as I said earlier my device uploads all the data to the server and there in i wanted to automate the decoding process of the logs collected. I did not want to use the trice application for the decoding process to reduce complexity

**Reply by watersb1323** [reply]
Apologies, I don't fully understand. Are you saying that you are now successfully decoding the data without using the trice application? If so, are you using Python to do that?

**Reply by rokath** [reply]
When writing a trice decoder in Python,  you could start with a [TCOBSv1](https://github.com/rokath/tcobs/blob/master/TCOBSv1/docs/TCOBSv1Specification.md#tcobsv1-specification) decoder in Python with Tests because the TREX format uses TCOBSv1 for framing combined with compression.

**Reply by rokath** [reply]
Have a look in the Go code https://github.com/rokath/tcobs/blob/master/TCOBSv1/tcobsDecode.go to get a first implementation idea for Python.

**Reply by rokath** [reply]
Just found this: https://www.ardanlabs.com/blog/2020/07/extending-python-with-go.html

**Reply by heezes** [reply]
@watersb1323 That was the idea but never really went ahead with writing the decoding method in python

**Reply by rokath** [reply]
Using COBS instead of TCOBS in the taget code would be possible as option. See #337. It will increase the raw trice byte stream maybe by about guessed 15% but there is a Python COBS decoder out of the shelf. Why on earth is Python so important after Go is in the world? And why it is not attraktive to use the trice binary for decoding and fetch the text logs via TCP4 or a file interface. The trice tool can send as several instances text logs (`trice log -port com1 -ds`) to a single display server `trice ds`.  Maybe it is worth to simply implement the `trice ds` functionality in Python?

**Reply by watersb1323** [reply]
My specific use case is the following:

1. Reduce code size
2. Send logs from MCU to PC to log to file
3. Send metrics from MCU to PC and transmit to MQTT broker before logging to file

We are running QNX OS on PC. No support for Go, hence looking to C++/Python option

**Reply by rokath** [reply]
Just to be sure, I understood right: You intend to use the `TRICE` macros on an embedded MCU board and to decode the binary trice stream on an embedded PC running QNX, probably to send the messages to an MQTT broker, right? *I never thought about such a use case.* I recommend to use **C** for decoding. [TCOBSv1](https://github.com/rokath/tcobs#4-current-state) offers a well tested C-decoder. The [TREX](https://github.com/rokath/trice/blob/refactor/docs/TriceV1.0Specification.md#62-main-stream-logs) format (This is the `refactor` branch!) will probably stay stable and is straight forward to parse: After TCOBSv1 decoding, the first 2 bytes contain in the 3 MSB a selector and the 13-bit ID. The selector tells about the (time-) stamp byte count, you can skip as a first approach, as well as the C(ycle)-byte, and the N-byte gives you the length of the following data bytes, what should be the rest of the decoded buffer. You could parse [til.json](https://github.com/rokath/trice/blob/refactor/til.json) with a C-lib but I recommend to write a small helper program (may be in Phython or **Go** :-) to convert the JSON file in a C-struct during compile time, which you include in your C-decoder source. Example:
```json
	"1014": {
		"Type": "TRICE",
		"Strg": "sig:Runtime generated strings\\n"
	},
	"1030": {
		"Type": "TRICE8_8",
		"Strg": "tst:TRICE8_8 %d %d %d %d %d %d %d %d\\n"
	},
```
to
```c
// ID, bit-width, value-count, format-string
{ 1014, 32, 0, "sig:Runtime generated strings\n" },
{ 1030, 8, 8, "tst:TRICE8_8 %d %d %d %d %d %d %d %d\n" },
```
Using this struct, you can than take the 13-bit ID to find the appropriate format string `pFmt`. With the bit-width information you can parse the data bytes and assign them to values v[12] and then you can call `sprintf( buf, pFmt, v[0], ... [v11])`. 
This would be a nice **trice** extension as a separate project I could link to or as a contribution as **trice** project extension.

Would it be an option to send the raw byte stream to the MQTT broker and to do the decoding there?

**Reply by rokath** [reply]
Now added #338 as a to do.

**Reply by rokath** [reply]
https://github.com/rokath/trice/issues/338 done in branch refactor.

**Reply by watersb1323** [reply]
Thanks for that explanation. That sounds like a useful implementation.

What I want to do though is to parse the stream, extract and transmit some but not all TRICE messages received from the MCU as they are received in the initial decoding and parsing process.

The idea is to use a specific string identifier to identify metrics on the MCU and then the parser will decide whether to:

A. Publish the string to the log file
B. Publish the string to the log file and also send this information to the MQTT broker as a metric, e.g. MCU temperature

Thus, I would rather do this in Python/C++ as it will be easier to handle the MQTT comms. ![image-3.png](https://user-images.githubusercontent.com/25089463/203067184-ce9d5f18-2a5d-462c-8b5e-311a7dacc033.png)

**Reply by watersb1323** [reply]
To your point on sending the raw byte stream to the MQTT broker, I would rather keep that decentralized so it can scale.

**Reply by rokath** [reply]
Selecting some of the messages is possible using channel specific ID ranges. See #329. This will come in a few weeks. Right now you already could generate IDs only between 2000 and 5999 and give "mqtt:..." messages manually the IDs 100...499 for example. See [trice-id-management-hints](https://github.com/rokath/trice/blob/refactor/docs/TriceIDManagement.md#trice-id-management-hints) (maybe partially outdated). Adding an mqtt channel is to be done by extending []colorChannels in [lineTransformerANSI.go](https://github.com/rokath/trice/blob/refactor/internal/emitter/lineTransformerANSI.go). For decoding help see [til.json.c](https://github.com/rokath/trice/blob/refactor/til.json.c).

**Reply by rokath** [reply]
You could consider decoding only the mqtt trice messages and to write the log file as binary logfile. With the **trice** tool you can view binary logfiles.

**Reply by watersb1323** [reply]
Good idea, I will add the MQTT channel. I think outside of this, the most practical solution will be to use the off-the-shelf COBS decoder with Python to process the data so I can intercept and transmit the MQTT channel messages. 

I can implement post-processing with existing TRICE app for now until [#337](https://github.com/rokath/trice/issues/337) is available.

**Reply by rokath** [reply]
@watersb1323, you can go ahead now using branch refactor. The information ./docs/TriceUserGuide.md should be sufficient. If you give me the needed syntax I could extend the trice tool to generate a `til.json.py` for you.

**Reply by Rafi22222** [reply]
Hi Thomas
Regarding your answer from Mar 6, 2022 , about integration of the logger with :
"Read https://github.com/rokath/trice/blob/master/docs/TriceMessagesEncoding.md and https://pythonhosted.org/cobs/cobs.cobs.html.
Write a Python program to read til.json"
...

After 4 years, and many new features, is there another easier recommended way integrate the 
trice (in logger mode) with existing application ?
Thanks
Rafi.

**Comment by rokath** [comment, marked as answer]
I added an issue: #265

**Comment by rokath** [comment]
Next release (v0.50.0) has a new switch `-tcp`. Also have a look in the user manual.

**Comment by rokath** [comment]
Right now we still have no solution for this. But it should be not too complicated to implement that. This is an Open Source project, so anybody needing such a solution can implement it. It is not complicated, especially it is a separate implementation and does not interfere with the main project. Some ideas:

- Configure Trice package framing firstly to COBS. There is ready-to-use code in C and Python for decoding.
- In Python, I guess, exist JSON parser.
- In C, if parsing the JSON Trice ID list is an issue, you can use `trice generate ...`to get C code like this:

```C
//! \file til.c
//! ///////////////////////////////////////////////////////////////////////////

//! Trice generated code - do not edit!

#include "til.h"

//! triceFormatStringList contains all trice format strings together with id and parameter information.
//!
//! The bitWidth value is not transmitted in the binary data stream and needed for its decoding.
//! The paramCount is de-facto not needed. It is derivable from the received data, see docs/TriceUserManual.md#binary-encoding.
//! It is recommended to check if both values are matching. A negative paramCount indicates, that its value is unknown at compile time.
const triceFormatStringList_t triceFormatStringList[] = {
	/* Trice type (  extended  ) */  //  id, bitWidth, paramCount, format-string
	/*      trice (  trice32_9 ) */ { 14016,  32,  9, "rd:trice %d, %d, %d, %d, %d, %d, %d, %d, %d\n" },
	/*      trice (     trice0 ) */ { 14224,  32,  0, "\n" },
	/*    trice32 (  trice32_1 ) */ { 14337,  32,  1, "msg:%u (%%u)\n" },
	/*     TRICE8 (  TRICE8_10 ) */ { 15063,   8, 10, "rd:TRICE8 %d, %d, %d, %d, %d, %d, %d, %d, %d, %d\n" },
	/*     Trice8 (   Trice8_9 ) */ { 15124,   8,  9, "rd:Trice8 %d, %d, %d, %d, %d, %d, %d, %d, %d\n" },
	/*     TRICE8 (   TRICE8_5 ) */ { 15058,   8,  5, "rd:TRICE8 %d, %d, %d, %d, %d\n" },
	/*      TRice (     TRice0 ) */ { 14885,  32,  0, "TEST:yellow+h:black\n" },
	/*    Trice64 (  Trice64_1 ) */ { 15560,  64,  1, "rd:Trice64 %d\n" },
	/*      trice (  trice32_1 ) */ { 15860,  32,  1, "rd:TRICE float %9.f (%%9.f)\n" },
...
	/*  TRICE64_0 (  TRICE64_0 ) */ { 16157,  64,  0, "w: Hello! 👋🙂 \a\n" },
	/*      TRICE (     TRICE0 ) */ { 14658,  32,  0, "interrupt:magenta+i:default+h\n" },
};

//! triceFormatStringListElements holds the compile time computed count of list elements.
const unsigned triceFormatStringListElements = sizeof(triceFormatStringList) / sizeof(triceFormatStringList_t);
```

- The main task to solve, is to write a parser in C or Python for the binary (COBS-decoded) Trice data to get the Trice IDs and the payload.

**Comment by Rafi22222** [comment]
Hi Thomas
Thank you very much for all the helpfull and informative answers!
BR
Rafi.

קבל ‏Outlook עבור Android‏<https://aka.ms/AAb9ysg>
________________________________
